Add Browse Catalog to Project context view.

jeff-phillips-18 opened the pull request:
![image](https://user-images.githubusercontent.com/11633780/33575694-7c9514b6-d90b-11e7-9e28-6d82d42c83eb.png)

Waiting on final placement and visuals from @jennyhaines for the Catalog nav item.

@openshift/team-ux-review 

/hold

spadgett commented on app/scripts/app.js:
```diff
@@ -61,6 +61,10 @@ angular
         templateUrl: 'views/create-project.html',
         controller: 'CreateProjectController'
       })
+      .when('/project/:project/project-browse-catalog', {
```
Personally, I like the shorter `/project/:project/catalog`

spadgett commented on app/scripts/controllers/overview.js:
```diff
@@ -1090,6 +1092,10 @@ function OverviewController($scope,
   });
 
+  $scope.browseCatalog = function() {
+    $location.url('project/' + $scope.projectName + '/project-browse-catalog');
```
Suggest adding this to the `Navigate` service so we only ever have one place where we build the URL.

spadgett commented on app/scripts/controllers/projectBrowseCatalog.js:
```diff
@@ -0,0 +1,34 @@
+             AuthorizationService,
+             AuthService,
+             CatalogService) {
+      AuthService.withUser().then(function() {
```
We should really be calling `ProjectsService.get()` to make sure the project exists. In that case, you don't need `AuthService.withUser()`.

spadgett commented on app/scripts/controllers/projectBrowseCatalog.js:
```diff
@@ -0,0 +1,34 @@
+          }
+
+          CatalogService.getCatalogItems().then(function (items) {
+            $scope.catalogItems = items;
```
It would be nice to show the image streams and templates from the current project as well since we have a project context.

jeff-phillips-18 replied:
We don't currently support browsing and choosing image streams. I can add project templates though.

spadgett replied:
> We don't currently support browsing and choosing image streams.

We support builder images in the catalog itself.

rhamilto commented on app/styles/_project-browse-catalog.less:
```diff
@@ -0,0 +1,37 @@
+.project-browse-catalog {
+  .services-view-container {
+    background: @color-pf-black-200;
+    min-height: calc(100vh ~"-" 101px);
```
We should avoid using vh units since they aren't reliable on mobile devices.

jeff-phillips-18 commented:
Addressed all issues except for showing the image streams and templates from the current project.

Updated the nav item icon to the pficon-service-catalog per latest designs.

![image](https://user-images.githubusercontent.com/11633780/33664073-577f113a-da60-11e7-83dd-0b1067d642fb.png)

spadgett commented:
> Addressed all issues except for showing the image streams and templates from the current project.

We can leave that as a follow on.

spadgett commented on app/views/directives/header/header.html:
```diff
@@ -40,7 +40,7 @@
       <li ng-if-end role="menuitem"><a ng-href="project/{{currentProjectName}}/create?tab=fromFile">Import YAML / JSON</a></li>
 
-      <li ng-if-start="catalogLandingPageEnabled" role="menuitem"><a href="./">Browse Catalog</a></li>
+      <li ng-if-start="catalogLandingPageEnabled" role="menuitem"><a href="project/{{currentProjectName}}/catalog">Browse Catalog</a></li>
```
Ideally we'd add a filter to build the URL using the `Navigate` service, similar to

https://github.com/openshift/origin-web-console/blob/master/app/scripts/filters/resources.js#L469

jeff-phillips-18 commented:
Added project templates and image streams to the list of items being browsed.

jeff-phillips-18 commented:
Set the icon back to fa-book for the catalog item on the nav menu.

jeff-phillips-18 commented:
That completes the UI visuals updates per @jennyhaines

Still WIP for https://github.com/openshift/origin-web-catalog/pull/587

@jwforres Are we OK with having 'Browse Catalog' on the 'Add to Project' dropdown still? It does the same thing as clicking on the nav item. While it is redundant, it may help the novice. Easily removable if we want to avoid the redundancy.

jwforres commented:
Yeah I wouldn't remove it from add to project, our existing users expect it
there.

On Dec 8, 2017 8:29 AM, "Jeffrey Phillips" <notifications@github.com> wrote:

> That completes the UI visuals updates per @jennyhaines
> <https://github.com/jennyhaines>
>
> Still WIP for openshift/origin-web-catalog#587
> <https://github.com/openshift/origin-web-catalog/pull/587>
>
> @jwforres <https://github.com/jwforres> Are we OK with having 'Browse
> Catalog' on the 'Add to Project' dropdown still? It does the same thing as
> clicking on the nav item. While it is redundant, it may help the novice.
> Easily removable if we want to avoid the redundancy.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/2563#issuecomment-350263216>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7ZA2ZgvGfNfD3JyOAV4pEkAsmsIeks5s-TnHgaJpZM4Q1Tnc>
> .
>

spadgett commented:
@jeff-phillips-18 I merged openshift/origin-web-catalog#587. We should be ready to cut a v3.9.0-alpha.0 release of catalog.

jeff-phillips-18 commented:
openshift/origin-web-catalog#587 is available, removed WIP

/hold cancel

spadgett commented on app/scripts/controllers/projectBrowseCatalog.js:
```diff
@@ -0,0 +1,49 @@
+          // Show all catalog items, project templates, and project image streams
+          CatalogService.getCatalogItems().then(function (items) {
+            Catalog.getProjectCatalogItems(currentProjectName, false, true).then( _.spread(function(catalogServiceItems, errorMessage) {
+              DataService.list(imageStreamsVersion, {namespace: currentProjectName}, function(isData) {
```
Pass `true` for `includeImages` to `getProjectCatalogItems` instead?

https://github.com/openshift/origin-web-catalog/blob/master/src/services/catalog.service.ts#L116-L125

jeff-phillips-18 replied:
The CatalogService passes true to Catalog if i!sTemplateServiceBrokerEnabled

spadgett replied:
> The CatalogService passes true to Catalog if i!sTemplateServiceBrokerEnabled

I think you might be looking at a different method / argument? Unless I'm misreading the code, there's one for image streams for the `getProjectCatalogItems` method you're calling directly. You're passing `false` just above.

jeff-phillips-18 replied:
Yeah, I was looking at the CatalogService.getCatalogItems call.

However, the Catalog.getProjectCatalogItems defaults the includeImages and includeTemplates to true:

`  public getProjectCatalogItems(projectName: string, includeImages: boolean = true, includeTemplates: boolean = true, partialObjectMetadataList: boolean = false ) {
`

so I think this is OK, right? When testing I do see both project templates and project image streams when browsing in project context which I do not see from the landing page.

spadgett commented on app/scripts/filters/resources.js:
```diff
@@ -475,6 +475,15 @@ angular.module('openshiftConsole')
+        Navigate.catalogURL(project) :
+        angular.isObject(project) ?
+          Navigate.catalogURL(project.metadata && project.metadata.name):
+          Navigate.catalogURL(''); // fail case will invoke default route (projects list)
```
> fail case will invoke default route (projects list)

Looks like that's not the case looking at the implementation below.

spadgett replied:
I realize this was copied from the overview method, but I think something like this is better:

```js
if (!_.isString(project)) {
  project = _.get(project, 'metadata.name');
}
return Navigate.catalogURL(project);
```

Or possibly move that logic into `Navigate.catalogURL`

jeff-phillips-18 replied:
I've moved it into Navigate.catalogURL

spadgett commented on app/scripts/services/navigate.js:
```diff
@@ -82,6 +82,17 @@ angular.module("openshiftConsole")
       },
 
+      catalogURL: function(projectName) {
+        return "project/" + encodeURIComponent(projectName) + "/catalog";
```
Maybe

```js
if (!projectName) {
  return "./";
}

return "project/" + encodeURIComponent(projectName) + "/catalog";
```

spadgett commented on app/scripts/controllers/projectBrowseCatalog.js:
```diff
@@ -0,0 +1,49 @@
+          // Show all catalog items, project templates, and project image streams
+          CatalogService.getCatalogItems().then(function (items) {
+            Catalog.getProjectCatalogItems(currentProjectName).then( _.spread(function(catalogServiceItems, errorMessage) {
+              DataService.list(imageStreamsVersion, {namespace: currentProjectName}, function(isData) {
```
I think we can remove this call? Aren't we fetching it twice now since `getProjectCatalogItems` also request image streams?

spadgett replied:
nit: We could request these in parallel and use `$q.all` so the page load slightly faster.

spadgett commented on app/scripts/filters/resources.js:
```diff
@@ -475,6 +475,11 @@ angular.module('openshiftConsole')
+    return function(project) {
+      return Navigate.catalogURL(project);
+    };
+  })
```
nit: this is fine, but I think it could just be

```js
.filter('catalogURL', function(Navigate) {
  return Navigate.catalogURL;
})
```

spadgett commented on app/scripts/services/navigate.js:
```diff
@@ -82,6 +82,22 @@ angular.module("openshiftConsole")
+        return "project/" + encodeURIComponent(projectName) + "/catalog";
+      },
+
+      toProjectCatalog: function(projectName, search) {
```
nit: You might change the parameter name to `project` since it can be an object, too. (Probably worth a comment as well.)

jeff-phillips-18 replied:
renamed and added comments

serenamarie125 commented:
I'd suggest the icon to be set back to pficon-service-catalog, this way, when PatternFly updates the icon, OpenShift will not need to update the code to get the new icon.

serenamarie125 commented on app/scripts/constants.js:
```diff
@@ -412,6 +412,14 @@ angular.extend(window.OPENSHIFT_CONSTANTS, {
+    },
+    {
+      label: "Catalog",
+      iconClass: "fa fa-book",
```
I'd suggest the icon to be set back to pficon-service-catalog, this way, when PatternFly updates the icon, OpenShift will not need to update the code to get the new icon.  @jeff-phillips-18

jeff-phillips-18 commented:
Thanks for you diligence @spadgett 

I've addressed those issues.

spadgett commented:
/lgtm
